### PR TITLE
Support generalized "all fields reference"

### PR DESCRIPTION
**Pull request by kasiafi**
refers to https://github.com/prestosql/presto/issues/875

### CODE REVIEW

**Comment by kasiafi** [comment]
A couple of tests in `TestSqlParserErrorHandling` fail. `Actual` error messages are broken. I need help with resolving it.

File: presto-parser/src/main/java/io/prestosql/sql/SqlFormatter.java [hunk 1]
```diff
@@ -393,7 +394,21 @@ protected Void visitSingleColumn(SingleColumn node, Integer indent)
         @Override
         protected Void visitAllColumns(AllColumns node, Integer context)
         {
-            builder.append(node.toString());
+            if (node.getTarget().isPresent()) {
```
Comment (electrum) [inline presto-parser/src/main/java/io/prestosql/sql/SqlFormatter.java]: This could simplify to
```java
node.getTarget().ifPresent(value -> builder
        .append(formatExpression(value, parameters))
        .append("."));
builder.append("*");

File: presto-parser/src/main/java/io/prestosql/sql/tree/AllColumns.java [hunk 2]
```diff
@@ -82,23 +93,33 @@ public boolean equals(Object o)
...
+        return Objects.hash(aliases, target);
     }
 
     @Override
     public String toString()
     {
-        if (prefix.isPresent()) {
-            return prefix.get() + ".*";
+        StringBuilder builder = new StringBuilder();
+        if (target.isPresent()) {
```
Comment (electrum) [inline presto-parser/src/main/java/io/prestosql/sql/tree/AllColumns.java]: This could simplify to
```java
target.ifPresent(value -> builder
        .append(value.toString())
        .append("."));
builder.append("*");

File: presto-parser/src/main/java/io/prestosql/sql/tree/DefaultTraversalVisitor.java [hunk 3]
```diff
@@ -134,6 +134,16 @@ protected R visitSelect(Select node, C context)
         return null;
     }
 
+    @Override
+    protected R visitAllColumns(AllColumns node, C context)
+    {
+        if (node.getTarget().isPresent()) {
```
Comment (electrum) [inline presto-parser/src/main/java/io/prestosql/sql/tree/DefaultTraversalVisitor.java]: Use `ifPresent`
```java
node.getTarget().ifPresent(value -> process(value, context));

**Comment by electrum** [comment]
@martint will probably have a better idea on how to fix the error reporting, but I was able to fix it by rewriting the rule to avoid duplicate paths for `expression`:
```
selectItem
    : ( expression
        ( (AS? identifier)?
        | '.' ASTERISK (AS? columnAliases)?
        )
      | ASTERISK
      )
    ;

**Comment by kasiafi** [comment]
I tried to change the grammar the way you suggested, but it required adding context labels and I couldn't fit them in. I wonder why this change breaks error messages in context independent of selectItem.

**Comment by electrum** [comment]
You don't need context labels. Instead, change `AstBuilder` to implement `visitSelectItem`. It is ugly but should work.

**Comment by kasiafi** [comment]
After changing the grammar and AstBuilder, there’s still a problem of the same kind.
```
SELECT a AS z FROM t WHERE a.

Expected :line 1:31: mismatched input '<EOF>'. Expecting: <identifier>
Actual   :line 1:29: mismatched input '.'. Expecting: '%', '*', '+', '-', '/', 'AT', '||', <expression>
```

**Comment by electrum** [comment]
This error message change seems legitimate now that field reference is an expression. So we should update the expected to match.

**Comment by kasiafi** [comment]
Applied.

**Comment by kasiafi** [comment]
Ready for review.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -1808,53 +1809,151 @@ private Scope computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope so
...
+            }
+            if (!allColumns.getAliases().isEmpty()) {
+                validateColumnAliases(allColumns.getAliases(), ((RowType) type).getFields().size());
+            }
+            analysis.recordSubqueries(node, expressionAnalysis);
+
+            for (int i = 0; i < ((RowType) type).getFields().size(); i++) {
+                Expression outputExpression = new SubscriptExpression(expression, new LongLiteral("" + (i + 1)));
+                ExpressionAnalysis outputExpressionAnalysis = analyzeExpression(outputExpression, scope);
+                outputExpressionBuilder.add(outputExpression);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: This can be a problem for non-deterministic or expensive expressions, since it's inlining the expression into each of the expanded select items. This should be handled during planning by adding projection that dereferences each field.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 5]
```diff
@@ -1808,53 +1809,151 @@ private Scope computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope so
...
+
+                Type outputExpressionType = outputExpressionAnalysis.getType(outputExpression);
+                if (node.getSelect().isDistinct() && !(outputExpressionType.isComparable())) {
+                    throw new SemanticException(TYPE_MISMATCH, node.getSelect(), "DISTINCT can only be applied to comparable types (actual: %s)", type.getTypeParameters().get(i));
+                }
+
+                Optional<String> oldName = ((RowType) type).getFields().get(i).getName();
+                if (!allColumns.getAliases().isEmpty()) {
+                    String name = (allColumns.getAliases().get(i)).getValue();
+                    itemOutputFieldBuilder.add(Field.newUnqualified(Optional.of(name), outputExpressionType, Optional.empty(), oldName, true));
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: I don't think using `oldName` here is correct. The `originName` in `Field` is used for column access control checks, so it has to be a proper column in the source table.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -1808,53 +1809,151 @@ private Scope computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope so
...
-                            throw new SemanticException(MISSING_TABLE, item, "Table '%s' not found", starPrefix.get());
-                        }
-                        if (!node.getFrom().isPresent()) {
-                            throw new SemanticException(WILDCARD_WITHOUT_FROM, item, "SELECT * not allowed in queries without FROM clause");
+                    List<Field> fields = relationType.resolveFieldsWithPrefix(Optional.of(prefix));
+
+                    if (!fields.isEmpty()) {
+                        // successfully dereferenced a table
+                        if (!allColumns.getAliases().isEmpty()) {
+                            throw new SemanticException(INVALID_PARAMETER_USAGE, node.getSelect(), "SELECT t.* FROM table does not allow column aliases");
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: `INVALID_PARAMETER_USAGE` doesn't seem like the right error code for this. We probably need a new one to handle this scenario.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: Yes, the expression is inlined in each expanded select item, but only 'syntactically'. The Planner identifies all occurrences as the same expression (via `TranslationMap`), and it is executed once.

Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: I ran `EXPLAIN` on
`SELECT (SELECT (min(regionkey), max(regionkey), true) FROM nation).*;`.
It only tablescaned once, computed the ROW from the subquery and then there was a Project which dereferenced each field. I think it is the plan you wanted.

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java [hunk 7]
```diff
@@ -112,7 +112,7 @@
                 {"SELECT a AS z FROM t WHERE x = 1 + ",
                         "line 1:36: mismatched input '<EOF>'. Expecting: <expression>"},
                 {"SELECT a AS z FROM t WHERE a. ",
-                        "line 1:31: mismatched input '<EOF>'. Expecting: <identifier>"},
+                        "line 1:29: mismatched input '.'. Expecting: '%', '*', '+', '-', '/', 'AT', '||', <expression>"},
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: I can't explain this change. Surprisingly, it's a change in the behavior of the Parser (who provides the location of the error), not the ErrorHandler.

**Comment by kasiafi** [comment]
Updated, added some more tests.

**Comment by martint** [comment]
Squash `Test generalized AllColumns` into `Implement generalized AllColumns in StatementAnalyzer`

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -1879,89 +1875,180 @@ private Scope computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope so
...
+                if (expression instanceof DereferenceExpression) {
+                    prefix = DereferenceExpression.getQualifiedName((DereferenceExpression) expression);
+                }
+                else if (expression instanceof Identifier) {
+                    prefix = QualifiedName.of(ImmutableList.of((Identifier) expression));
+                }
+                if (prefix != null) {
+                    // try match prefix to a table
+                    RelationType relationType = scope.getRelationType();
+                    List<Field> fields = relationType.resolveFieldsWithPrefix(Optional.of(prefix));
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: This should fail but does not (use memory connector to try it):

```
CREATE SCHEMA a;
USE a;
CREATE TABLE b (g varchar);
CREATE TABLE t (x row(f1 bigint, f2 bigint));

SELECT a.b.* FROM a.b, t AS a;  --- should fail but does not
```

(BTW, this might be a pre-existing bug).

File: presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java [hunk 9]
```diff
@@ -999,39 +999,71 @@ public void testCountColumn()
...
+        assertQuery("SELECT orders.* FROM orders");
     }
 
     @Test
-    public void testQualifiedWildcard()
+    public void testSelectAllFromRow()
     {
-        assertQuery("SELECT orders.* FROM orders");
+        // named row fields and aliases
+        assertQuery("SELECT CAST((ROW (1, 'a')) AS ROW (f1 bigint, f2 varchar(1))).* AS (alias1, alias2)", "SELECT 1 alias1, 'a' alias2");
```
Comment (martint) [inline presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java]: Move these tests to something under `io.prestosql.sql.query` so that they don't run with every connector (given that they don't need to access tables)

File: presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java [hunk 10]
```diff
@@ -253,10 +256,42 @@ public void testHavingReferencesOutputAlias()
...
+                .hasErrorCode(COLUMN_NOT_FOUND);
+
+        // wildcard with invalid prefix
         assertFails("SELECT foo.* FROM t1")
                 .hasErrorCode(TABLE_NOT_FOUND);
+
+        assertFails("SELECT a.b.c.d.* FROM t1")
+                .hasErrorCode(TABLE_NOT_FOUND);
+
+        // name aliases are not allowed when selecting all columns from a table
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java]: This should say "not currently supported"? It should be allowed from the `If N ≤ 3 and PIC_N is equivalent to an exposed <table or query name> that is in scope` case. It's actually one of the best uses for this feature. If you want to rename all the columns from a table you currently have to do `SELECT c1 AS n1, c2 as n2, c3 as n3, ...`. With this, it can be simplified for `SELECT t.* AS (n1, n2, n3)` and save some typing.

File: presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java [hunk 11]
```diff
@@ -253,10 +256,42 @@ public void testHavingReferencesOutputAlias()
...
+        assertFails("SELECT (1, 2).* AS (a, a) FROM t1")
+                .hasErrorCode(DUPLICATE_COLUMN_NAME);
+
+        // wildcard with no RowType expression
+        assertFails("SELECT non_row.* FROM (VALUES ('true', 1)) t(non_row, b)")
+                .hasErrorCode(TABLE_NOT_FOUND);
+
+        // wildcard with no RowType expression nested in a row
+        assertFails("SELECT t.row.non_row.* FROM (VALUES (CAST(ROW('true') AS ROW(non_row boolean)), 1)) t(row, b)")
+                .hasErrorCode(INVALID_ALL_COLUMNS_TARGET);
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java]: This should probably be a TYPE_MISMATCH error. Basically, it expected a ROW type but found something else.

File: presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java [hunk 12]
```diff
@@ -253,10 +256,42 @@ public void testHavingReferencesOutputAlias()
...
+        // name aliases are not allowed when selecting all columns from a table
+        assertFails("SELECT t.* AS (a, b, c) FROM (VALUES (1, 2, 3), (10, 20, 30), (100, 200, 300)) t")
+                .hasErrorCode(NOT_SUPPORTED);
+
+        // aliases mismatch
+        assertFails("SELECT (1, 2).* AS (a) FROM t1")
+                .hasErrorCode(MISMATCHED_COLUMN_ALIASES);
+
+        // duplicate alias
+        assertFails("SELECT (1, 2).* AS (a, a) FROM t1")
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java]: This should be allowed, just like `SELECT 1 AS a, 2 AS a` is allowed. The ambiguity should be reported when someone tries to de-reference `a`. For instance:

```sql
SELECT a FROM (
    SELECT (1, 2).* AS (a,a) 
    FROM t
) u
```

File: presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java [hunk 13]
```diff
@@ -999,39 +999,71 @@ public void testCountColumn()
...
+
+        // unnamed row fields and aliases
+        assertQuery("SELECT (ROW (1, 'a')).* AS (alias1, alias2)", "SELECT 1 alias1, 'a' alias2");
```
Comment (martint) [inline presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java]: If I'm not mistaken, `assertQuery` doesn't validate that the output columns are named the same. So, `SELECT 1 alias1, 'a' alias2` is not testing what you think. I'm not sure we even have a way in the current framework to do so.

Maybe replace it with something like this:

```sql
SELECT alias1, alias2 
FROM (
    SELECT (ROW (1, 'a')).* AS (alias1, alias2)
) t
```

Then we can be sure that the aliasing is working fine.

File: presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java [hunk 14]
```diff
@@ -999,39 +999,71 @@ public void testCountColumn()
...
+        assertQuery("SELECT (SELECT (name, regionkey) FROM nation WHERE name='ALGERIA').*", "SELECT 'ALGERIA', 0");
+        assertQuery("SELECT (SELECT (count(*), true) FROM nation WHERE regionkey = 0).*", "SELECT 5, true");
+
+        // multiple wildcards
+        assertQuery("SELECT ROW(a, b).*, ROW(b, c).*, ROW(a, c).* FROM (VALUES (1, 2, 3)) t(a, b, c)", "VALUES (1, 2, 2, 3, 1, 3)");
+
+        // non-deterministic expression precomputed
+        MaterializedResult materializedResult = computeActual("SELECT (x, x, x, x, x, x, x, x).* FROM (SELECT rand()) T(x)");
+        long distinctValuesCount = materializedResult.getMaterializedRows().get(0).getFields().stream().distinct().count();
+        assertEquals(distinctValuesCount, 1, "rand() must be computed once only");
```
Comment (martint) [inline presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java]: Perfect!

File: presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java (hunk 10 above)
Comment (kasiafi) [reply presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java]: Apparently I got trapped (again) following the preexisting behavior.
Adding support for aliases when `IC` refers to a table.

**Comment by kasiafi** [comment]
applied comments

**Comment by kasiafi** [comment]
added tests for asterisked identifier chain analysis

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java [hunk 15]
```diff
@@ -99,6 +103,95 @@ public RelationType getRelationType()
         return relation;
     }
 
+    public Optional<AsteriskedIdentifierChainBasis> resolveAsteriskedIdentifierChainBasis(QualifiedName identifierChain)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java]: There's a lot of repetition for this logic and mutually recursive calls that make the code hard to follow. Here's an idea on how I would structure it:

```java
private boolean findLocally(Predicate<Scope> match)
{
    Scope scope = this;

    do {
        if (match.test(scope)) {
            return true;
        }

        Optional<Scope> parent = scope.getLocalParent();
        if (!parent.isPresent()) {
            break;
        }

        scope = parent.get();
    }
    while (true);

    return false;
}

public Optional<AsteriskedIdentifierChainBasis> resolveAsteriskedIdentifierChainBasis(QualifiedName identifierChain)
{
    int length = identifierChain.getParts().size();

    boolean foundTable = false;
    boolean foundField = false;

    if (length <= 3) {
        foundTable = findLocally(scope -> scope.getRelationType()
                .getAllFields().stream()
                .anyMatch(field -> field.getRelationAlias().isPresent() && field.getRelationAlias().get().equals(identifierChain));
    }

    if (length >= 2) {
        foundField = findLocally(scope -> scope.getRelationType()
                .getAllFields().stream()
                .anyMatch(field1 -> field1.getRelationAlias().isPresent() && field1.getName().isPresent()
                        && field1.getRelationAlias().get().equals(QualifiedName.of(identifierChain.getParts().get(0)))
                        && field1.getName().get().equals(identifierChain.getParts().get(1))
                        && field1.getType() instanceof RowType));
    }

    if (foundTable && foundField) {
        throw new PrestoException(AMBIGUOUS, ...);
    }

    if (foundTable) {
        return Optional.of(TABLE);
    }

    if (foundField) {
        return Optional.of(FIELD);
    }

    return getOuterQueryParent()
            .map(parent -> parent.resolveAsteriskedIdentifierChainBasis(identifierChain))
            .orElse(Optional.empty());
}
```

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java [hunk 16]
```diff
@@ -99,6 +103,95 @@ public RelationType getRelationType()
...
+        if (getOuterQueryParent().isPresent()) {
+            return getOuterQueryParent().get().resolveTableOrFieldBasis(identifierChain);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<AsteriskedIdentifierChainBasis> resolveTableBasisLocally(QualifiedName identifierChain)
+    {
+        if (relation.getAllFields().stream()
+                .anyMatch(field -> field.getRelationAlias().isPresent() && field.getRelationAlias().get().equals(identifierChain))) {
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java]: Actually, the spec says this about it:

Screen Shot 2019-10-16 at 10 40 01 PM

So that query should fail because there can't be two `t` range variables in the `FROM` clause.

[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java]: Here's another fun one that should be disallowed:

```
SELECT orders.* 
FROM orders, lineitem AS orders
```

[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java]: So such ambiguity should be sorted out when the `FROM` clause is analysed (that is before we start to analyse `SELECT` items).

[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java]: > To fix this, we'd need to record the available table and correlation names in the scope object. It's a much bigger change and probably out of scope. for this PR.

If we followed what the spec says about the `FROM` clause, there would be no ambiguous names in scope. However, I don't think we want to do it - it would break queries that are intuitively correct. And still the `WITH` clause might introduce ambiguity.

And if we want to deal with ambiguity of aliases/table names in scope, or even notice it, then we need to restructure `Scope` to track fields' origin instead of having a single `RelationType`.

[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java]: Even if we want to do it lazily, we need to record the individual aliases and table names. Otherwise, we can’t tell if there’s more than one at the point of resolution. Also, thinking about it more, RelationType is probably the wrong abstraction for how we need to do resolution anyway.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 17]
```diff
@@ -1885,89 +1885,213 @@ private void computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope sou
...
+        private void analyzeAllColumnsFromTable(
+                List<Field> fields,
+                AllColumns allColumns,
+                QuerySpecification node,
+                Scope scope,
+                ImmutableList.Builder<Expression> outputExpressionBuilder,
+                ImmutableList.Builder<SelectExpression> selectExpressionBuilder,
+                RelationType relationType)
+        {
+            if (!allColumns.getAliases().isEmpty()) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: This is repeated in both analysis methods. Pull it out into a separate method.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 18]
```diff
@@ -1885,89 +1885,213 @@ private void computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope sou
...
-                throw semanticException(NOT_SUPPORTED, allColumns, "Column aliases not supported");
+                        analyzeAllColumnsFromTable(fields, allColumns, node, scope, outputExpressionBuilder, selectExpressionBuilder, relationType.get());
+                    }
+                    else { // identifierChainBasis.get() == FIELD
+                        analyzeAllFieldsFromRowTypeExpression(expression, allColumns, node, scope, outputExpressionBuilder, selectExpressionBuilder);
+                    }
+                }
+                else {
+                    // analyze all cases of expression.* other than QualifiedName.*
+                    analyzeAllFieldsFromRowTypeExpression(expression, allColumns, node, scope, outputExpressionBuilder, selectExpressionBuilder);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: Instead of repeating this, I'd change the structure of the code slightly:

```java
if (allColumns.getTarget().isPresent()) {
    ...

    if (prefix != null) {
        ...

        if (identifierChainBasis.get() == TABLE) {
            ...
            analyzeAllColumnsFromTable(fields, allColumns, node, scope, outputExpressionBuilder, selectExpressionBuilder, relationType.get());
            return;
        }
    }
    analyzeAllFieldsFromRowTypeExpression(expression, allColumns, node, scope, outputExpressionBuilder, selectExpressionBuilder);
}
```

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 19]
```diff
@@ -1885,89 +1885,213 @@ private void computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope sou
...
+                            field.getType(),
+                            false,
+                            field.getOriginTable(),
+                            field.getOriginColumnName(),
+                            false));
+                }
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(fieldReference, scope);
 
-                Type type = expressionAnalysis.getType(expression);
+                Type type = expressionAnalysis.getType(fieldReference);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: Why not `Type type = field.getType()` :

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 20]
```diff
@@ -1885,89 +1885,213 @@ private void computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope sou
...
+                }
+                else {
+                    itemOutputFieldBuilder.add(new Field(
+                            field.getRelationAlias(),
+                            field.getName(),
+                            field.getType(),
+                            false,
+                            field.getOriginTable(),
+                            field.getOriginColumnName(),
+                            false));
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: This might need to be `field.isAliased()`. I don't recall what we use that flag for.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 21]
```diff
@@ -1885,89 +1885,213 @@ private void computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope sou
...
+            for (int i = 0; i < fields.size(); i++) {
+                Field field = fields.get(i);
                 int fieldIndex = relationType.indexOf(field);
-                FieldReference expression = new FieldReference(fieldIndex);
-                outputExpressionBuilder.add(expression);
-                ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, scope);
+                FieldReference fieldReference = new FieldReference(fieldIndex);
+                outputExpressionBuilder.add(fieldReference);
+                selectExpressionBuilder.add(new SelectExpression(fieldReference, Optional.empty()));
+                if (!allColumns.getAliases().isEmpty()) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: I'd structure this as:

```java
Optional<String> alias;
if (!allColumns.getAliases().isEmpty()) {
    alias = Optional.of((allColumns.getAliases().get(i)).getValue());
}
else {
    alias = field.getName();
}

itemOutputFieldBuilder.add(new Field(
        field.getRelationAlias(),
        alias,
        field.getType(),
        false,
        field.getOriginTable(),
        field.getOriginColumnName(),
        alias.isPresent()));
```

Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: Are you sure that `alias.isPresent()` is the right value for `aliased` parameter in the case  there were no column aliases?

Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: It is wrong actually. `TestDescribeOutput` fails.

**Comment by kasiafi** [comment]
Applied comments.

File: presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java [hunk 22]
```diff
@@ -361,6 +455,37 @@ private PlanBuilder project(PlanBuilder subPlan, Iterable<Expression> expression
...
+        for (int i = 0; i < selectExpressions.size(); i++) {
+            SymbolReference symbolReference = (SymbolReference) symbolReferences.get(i);
+            SelectExpression selectExpression = selectExpressions.get(i);
+            if (selectExpression.getTypeToUnfold().isPresent()) {
+                // expression to unfold of type Row: add field-reference expressions for all fields of the row
+                RowType type = selectExpression.getTypeToUnfold().get();
+                List<Type> unfoldedTypes = type.getTypeParameters();
+                for (int field = 0; field < unfoldedTypes.size(); field++) {
+                    unfoldingExpressions.add(new UnfoldingExpression(
+                            new SubscriptExpression(symbolReference, new LongLiteral("" + (field + 1))),
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java]: Why do we need to do this? We're already building the dereference expressions during analysis, so we should just have to keep track of them and pass them to the planner. Maybe SelectExpressions could be refactored to do that. 

I think we can also simplify how we add the projections are being added. Let me play around with the code a bit.

Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java]: Right! If we project the whole Row expression, and then add projections for field reference expressions on top of that, they should pick up the preprojected symbol for subscript base. That simplifies things a lot. Let me refactor that.

**Comment by kasiafi** [comment]
Applied comments. Refactored SelectExpression and simplified adding projections.

File: presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java [hunk 23]
```diff
@@ -162,29 +174,52 @@ public RelationPlan plan(QuerySpecification node)
...
+        List<Expression> outputs = expressions;
+        builder = handleSubqueries(builder, node, expressions);
+        List<Expression> outputExpressions = outputExpressions(selectExpressions);
 
         if (node.getOrderBy().isPresent()) {
             if (!analysis.isAggregation(node)) {
                 // ORDER BY requires both output and source fields to be visible if there are no aggregations
-                builder = project(builder, outputs, fromRelationPlan);
-                outputs = toSymbolReferences(computeOutputs(builder, outputs));
+                builder = project(builder, Iterables.concat(expressions, toSymbolReferences(fromRelationPlan.getFieldMappings())));
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java]: Can you add some tests containing `xxx.*` in queries involving aggregations and/or order by?

File: presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java [hunk 24]
```diff
@@ -203,6 +238,25 @@ public RelationPlan plan(QuerySpecification node)
...
     }
 
+    private boolean hasExpressionsToUnfold(List<SelectExpression> selectExpressions)
+    {
+        return selectExpressions.stream().map(SelectExpression::getUnfoldedExpressions).anyMatch(Optional::isPresent);
+    }
+
+    private List<Expression> outputExpressions(List<SelectExpression> selectExpressions)
+    {
+        ImmutableList.Builder<Expression> outputExpressionBuilder = ImmutableList.builder();
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java]: Just call this `result`. It's more concise without sacrificing readability.

File: presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java [hunk 25]
```diff
@@ -203,6 +238,25 @@ public RelationPlan plan(QuerySpecification node)
                 computeOutputs(builder, outputs));
     }
 
+    private boolean hasExpressionsToUnfold(List<SelectExpression> selectExpressions)
+    {
+        return selectExpressions.stream().map(SelectExpression::getUnfoldedExpressions).anyMatch(Optional::isPresent);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java]: Format across multiple lines:

```java
return selectExpressions.stream()
    .map(SelectExpression::getUnfoldedExpressions)
    .anyMatch(Optional::isPresent);
```

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java [hunk 26]
```diff
@@ -99,6 +105,93 @@ public RelationType getRelationType()
...
+            scopeForFieldReference = findLocally(scope -> scope.getRelationType()
+                    .getAllFields().stream()
+                    .anyMatch(field -> field.matchesPrefix(Optional.of(QualifiedName.of(identifierChain.getParts().get(0))))
+                            && field.getName().isPresent()
+                            && field.getName().get().equals(identifierChain.getParts().get(1))
+                            && field.getType() instanceof RowType));
+        }
+
+        if (scopeForTableReference.isPresent() && scopeForFieldReference.isPresent()) {
+            throw semanticException(AMBIGUOUS_IDENTIFIER_CHAIN, selectItem, "Reference '%s' is ambiguous", identifierChain);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java]: We should just use `AMBIGUOUS_NAME`, instead.
